src/core.lisp

```diff
@@ -1,3 +1,80 @@
+        (ab-perpendicular (vectors:vrot2 (vectors:v- (first triangle) (second triangle)) (/ pi 2))))
+    (and (< 0 (* (vectors:v. a-to-point bc-perpendicular)
+                 (vectors:v. b-to-point ca-perpendicular)))
+         (< 0 (* (vectors:v. b-to-point ca-perpendicular)
+                 (vectors:v. c-to-point ab-perpendicular))))))
+
+(defun calculate-depth (point2d point-a point-b point-c)
+  "Calculate depth of a point given depths of triangle vertices."
+  (let ((distance-a (vectors:v2norm (vectors:v- point2d
+                                                (first point-a))))
+        (distance-b (vectors:v2norm (vectors:v- point2d
+                                                (first point-b))))
+        (distance-c (vectors:v2norm (vectors:v- point2d
+                                                (first point-c)))))
+    (/ (+ (* (second point-a) distance-b distance-c)
+          (* (second point-b) distance-c distance-a)
+          (* (second point-c) distance-a distance-b))
+       (+ (* distance-a distance-b)
+          (* distance-b distance-c)
+          (* distance-c distance-a)))))
+
+(defun render (scene camera)
+  (let ((image (make-array (list (scene:camera-width camera) (scene:camera-height camera))
+                           :initial-element '(0 0 0)))
+        (depths (make-array (list (scene:camera-width camera) (scene:camera-height camera))
+                            :initial-element :infinity)))
+    (dolist (triangle (scene:scene-triangles scene))
+      (let* ((a-point (project-point-to-camera camera (first triangle)))
+             (b-point (project-point-to-camera camera (second triangle)))
+             (c-point (project-point-to-camera camera (third triangle)))
+             (minimal-x (max 0 (min (vectors:vx (first a-point))
+                                    (vectors:vx (first b-point))
+                                    (vectors:vx (first c-point)))))
+             (maximal-x (min (scene:camera-width camera)
+                             (max (vectors:vx (first a-point))
+                                  (vectors:vx (first b-point))
+                                  (vectors:vx (first c-point)))))
+             (minimal-y (max 0 (min (vectors:vy (first a-point))
+                                    (vectors:vy (first b-point))
+                                    (vectors:vy (first c-point)))))
+             (maximal-y (min (scene:camera-height camera)
+                             (max (vectors:vy (first a-point))
+                                  (vectors:vy (first b-point))
+                                  (vectors:vy (first c-point))))))
+        (loop for pixel-x from minimal-x to maximal-x
+              do (loop for pixel-y from minimal-y to maximal-y
+                       when (triangle-contains (list a-point b-point c-point) (list pixel-x pixel-y))
+                       do (let ((point-depth (calculate-depth
```

Comment:
There should be no offset in this case.
Or it could be better to place `(vectors:vec2 pixel-x pixel-y)` (first argument) at the same line as `calculate-depth` (function name).